src/gSP.cpp

```diff
@@ -2072,23 +2078,22 @@ struct ObjCoordinates
 			scaleH = 1.0f;
 		}
 
-		ulx = frameX;
-		uly = frameY;
-
 		uls = imageX;
 		ult = imageY;
-		lrs = uls + min(imageW, frameW * scaleW);
-		lrt = ult + min(imageH, frameH * scaleH);
+		lrs = uls + min(imageW, frameW * scaleW) - 1;
+		lrt = ult + min(imageH, frameH * scaleH) - 1;
 
-		// G_CYC_COPY does not allow texture filtering
+		// G_CYC_COPY (gSPBgRectCopy()) does not allow texture filtering
 		if (gDP.otherMode.cycleType != G_CYC_COPY) {
-			// Correct texture coordinates -0.5f if G_OBJRM_BILERP 
+			// Correct texture coordinates -0.5f and +0.5 if G_OBJRM_BILERP 
 			// bilinear interpolation is set
-			if ((gSP.objRendermode&G_OBJRM_BILERP) != 0) {
+			if ((gSP.objRendermode&G_OBJRM_BILERP) != 0 && 
+				(gDP.otherMode.textureFilter == 0x02) ||											// Kirby Crystal Shards
+				(gDP.otherMode.textureFilter == 0x00 && gSP.objRendermode&G_OBJRM_NOTXCLAMP != 0)) {// Worms Armageddon
```

Comment:
Two problems with this conditions:
1) Please add brackets to maker the condition more clear.
I guess, you mean
```
((gSP.objRendermode&G_OBJRM_BILERP)  != 0 &&  gDP.otherMode.textureFilter == 0x02)
 ||
(.otherMode.textureFilter == 0x00 && (gSP.objRendermode&G_OBJRM_NOTXCLAMP) != 0)
```
2) There are constants for texture filter modes:
gDP.otherMode.textureFilter == G_TF_BILERP instead of gDP.otherMode.textureFilter == 0x02
.otherMode.textureFilter == G_TF_POINT instead of gDP.otherMode.textureFilter == 0x00